File: apps/agent/src/agents/workspaceAgent/tools/searchMessages.ts
import { FunctionTool } from '@google/adk';
import { z } from 'zod';
import type { WorkspaceTokensStore } from '../../../storage/workspaceTokens.js';
import type { ExecFileLike } from '../gwsExec.js';
import { type RunGwsLogEvent, runGws } from '../runGws.js';
import { type InboxMessageSummary, LIST_INBOX_TOOL_NAME } from './listInbox.js';

/**
 * `search_messages` — Gmail search across the whole mailbox, not just the
 * inbox. Same shape as list_inbox: returns id+threadId+snippet summaries,
 * sub-agent calls get_message for full bodies on the matches it cares
 * about.
 */

export const SEARCH_MESSAGES_TOOL_NAME = 'search_messages';

export interface CreateSearchMessagesToolDeps {
  store: WorkspaceTokensStore;
  uid: string;
  execFile?: ExecFileLike;
  log?: (event: RunGwsLogEvent) => void;
}

export type SearchMessagesResult =
  | { status: 'ok'; messages: InboxMessageSummary[]; truncated?: boolean }
  | { status: 'error'; code: string; message: string };

const parameters = z.object({
  query: z
    .string()
    .min(1)
    .describe(
      'Gmail search syntax — e.g. "from:sarah newer_than:7d", "subject:invoice", "label:starred".',
    ),
  limit: z
    .number()
    .int()
    .positive()
    .max(50)
    .optional()
    .describe('Maximum number of messages to return (1–50). Default 10.'),
});

interface ListResponse {
  messages?: { id?: string; threadId?: string }[];
}

interface MessageMetadata {
  id?: string;
  threadId?: string;
  snippet?: string;
}

export function createSearchMessagesTool(deps: CreateSearchMessagesToolDeps): FunctionTool {
  const { store, uid, execFile, log } = deps;

  // biome-ignore lint/suspicious/noExplicitAny: zod nominal mismatch with ADK generics
  return new FunctionTool<any>({
    name: SEARCH_MESSAGES_TOOL_NAME,
    description:
      'Search Gmail across all labels/folders using Gmail query syntax. Returns id+threadId+snippet summaries — call get_message for full body. Read-only.',
    parameters,
    execute: async (input: unknown): Promise<SearchMessagesResult> => {
      const args = input as { query: string; limit?: number };
      const maxResults = args.limit ?? 10;

      const listResult = await runGws({
        store,
        uid,
        // Re-use the inbox list tool's name so all message-list logs cluster.
        toolName: LIST_INBOX_TOOL_NAME,
        service: 'gmail',
        resource: 'users.messages',
        method: 'list',
        params: { userId: 'me', q: args.query, maxResults },
        execFile,
        log,
      });
      if (listResult.status === 'error') {
        return { status: 'error', code: listResult.code, message: listResult.message };
      }
      const listBody = (listResult.body as ListResponse | null) ?? {};
      const ids = (listBody.messages ?? [])
        .map((m) => m.id)
        .filter((id): id is string => typeof id === 'string' && id.length > 0);
      if (ids.length === 0) {
        return { status: 'ok', messages: [] };
      }

      const detailResults = await Promise.all(
        ids.map((id) =>
          runGws({
            store,
            uid,
            toolName: SEARCH_MESSAGES_TOOL_NAME,
            service: 'gmail',
            resource: 'users.messages',
            method: 'get',
            params: { userId: 'me', id, format: 'metadata' },
            execFile,
            log,
          }),
        ),
      );
      const messages: InboxMessageSummary[] = [];
      for (const detail of detailResults) {
        if (detail.status !== 'ok') continue;
        const m = (detail.body as MessageMetadata | null) ?? {};
        if (!m.id) continue;
        messages.push({
          id: m.id,
          threadId: m.threadId ?? m.id,
          snippet: m.snippet ?? '',
        });
      }
      return { status: 'ok', messages };
    },
  });
}

File: apps/agent_py/src/lifecoach_agent/workspace_agent/tools/search_messages.py
"""`search_messages` — Gmail search across the whole mailbox, not just
the inbox. Returns id+threadId+snippet summaries; the sub-agent calls
get_messages for bulk full-body reads on matches it cares about.
"""

from __future__ import annotations

import asyncio
from typing import Any

from lifecoach_agent.workspace_agent.run_gws import RunGwsOk, run_gws
from lifecoach_agent.workspace_agent.tools._deps import WorkspaceToolDeps

SEARCH_MESSAGES_TOOL_NAME = "search_messages"


def create_search_messages_tool(deps: WorkspaceToolDeps) -> Any:
    async def search_messages(query: str, limit: int = 10) -> dict[str, Any]:
        """Search Gmail across all labels/folders using Gmail query syntax.
        Returns id+threadId+snippet summaries — call get_message for full
        body; use get_messages for bulk reads. Read-only.

        Args:
            query: Gmail search syntax — e.g. "from:sarah newer_than:7d",
                "subject:invoice", "label:starred".
            limit: Maximum number of messages to return (1–50). Default 10.
        """
        max_results = max(1, min(int(limit), 50))
        list_result = await run_gws(
            store=deps.store,
            uid=deps.uid,
            # Reuse list_inbox name so all message-list logs cluster.
            tool_name="list_inbox",
            service="gmail",
            resource="users.messages",
            method="list",
            params={"userId": "me", "q": query, "maxResults": max_results},
            build_client=deps.build_client,
            log=deps.log,
        )
        if not isinstance(list_result, RunGwsOk):
            return {"status": "error", "code": list_result.code, "message": list_result.message}

        body = list_result.body if isinstance(list_result.body, dict) else {}
        ids = [
            m.get("id")
            for m in (body.get("messages") or [])
            if isinstance(m, dict) and isinstance(m.get("id"), str)
        ]
        ids = [i for i in ids if i]
        if not ids:
            return {"status": "ok", "messages": []}

        details = await asyncio.gather(
            *[
                run_gws(
                    store=deps.store,
                    uid=deps.uid,
                    tool_name=SEARCH_MESSAGES_TOOL_NAME,
                    service="gmail",
                    resource="users.messages",
                    method="get",
                    params={"userId": "me", "id": mid, "format": "metadata"},
                    build_client=deps.build_client,
                    log=deps.log,
                )
                for mid in ids
            ]
        )
        messages: list[dict[str, Any]] = []
        for detail in details:
            if not isinstance(detail, RunGwsOk):
                continue
            m = detail.body if isinstance(detail.body, dict) else {}
            mid = m.get("id")
            if not isinstance(mid, str):
                continue
            messages.append(
                {
                    "id": mid,
                    "threadId": m.get("threadId") or mid,
                    "snippet": m.get("snippet") or "",
                }
            )
        return {"status": "ok", "messages": messages}

    from google.adk.tools.function_tool import FunctionTool

    search_messages.__name__ = SEARCH_MESSAGES_TOOL_NAME
    return FunctionTool(search_messages)
